### PR TITLE
Drag and drop support using left click(mouse 1)

### LINKED ISSUE #425: [Feature Request] Drag and drop a file to a specific window

### PR DESCRIPTION
**Pull request by Compro-Prasad**
Fixes #425

### CODE REVIEW

File: src/elisp/treemacs-mouse-interface.el [hunk 1]
```diff
@@ -29,6 +29,10 @@
 (require 'treemacs-filewatch-mode)
 (eval-and-compile (require 'treemacs-macros))
 
+(defvar treemacs-last-selected-button nil
```
Comment (Alexander-Miller) [inline src/elisp/treemacs-mouse-interface.el]: Use private double-dash.

File: src/elisp/treemacs-mouse-interface.el [hunk 2]
```diff
@@ -48,6 +52,17 @@ Must be bound to a mouse click, or EVENT will not be supplied."
        :no-error            t))
     (treemacs--evade-image)))
 
+(defun treemacs-downleftclick-action (event)
```
Comment (Alexander-Miller) [inline src/elisp/treemacs-mouse-interface.el]: AFAICT this method is largely a duplication of the leftclick-action above, only used for saving the clicked node. If so we should combine them.

File: src/elisp/treemacs-mouse-interface.el [hunk 3]
```diff
@@ -48,6 +52,17 @@ Must be bound to a mouse click, or EVENT will not be supplied."
...
+  "Select the window and record the node's path."
+  (interactive "e")
+  (when (eq 'down-mouse-1 (elt event 0))
+    (run-hooks 'mouse-leave-buffer-hook)
+    (select-window (elt (elt event 1) 0))
+    (goto-char (posn-point (cadr event)))
+    (setq treemacs-last-selected-button (treemacs-safe-button-get (treemacs-current-button) :path))
```
Comment (Alexander-Miller) [inline src/elisp/treemacs-mouse-interface.el]: The -safe-button-get is only needed when current-buffer and the button's buffer are different.

File: src/elisp/treemacs-mouse-interface.el [hunk 4]
```diff
@@ -95,6 +110,12 @@ Clicking on icons will expand a file's tags, just like
         (funcall (cdr (assoc state treemacs-doubleclick-actions-config)))))
     (treemacs--evade-image)))
 
+(defun treemacs-dragleftclick-action (event)
+  (interactive "e")
+  (when (and (eq 'drag-mouse-1 (elt event 0)) treemacs-last-selected-button)
+    (select-window (elt (elt event 2) 0))
```
Comment (Alexander-Miller) [inline src/elisp/treemacs-mouse-interface.el]: The path is not checked for being a file.

File: src/extra/treemacs-evil.el [hunk 5]
```diff
@@ -85,7 +85,8 @@ Then call ORIG-FUN with its ARGS and reopen treemacs if it was open before."
 (evil-define-key 'treemacs treemacs-mode-map (kbd "yy")     #'treemacs-copy-path-at-point)
 (evil-define-key 'treemacs treemacs-mode-map (kbd "yf")     #'treemacs-copy-file)
 (evil-define-key 'treemacs treemacs-mode-map (kbd "gr")     #'treemacs-refresh)
-(evil-define-key 'treemacs treemacs-mode-map [down-mouse-1] #'ignore)
```
Comment (Alexander-Miller) [inline src/extra/treemacs-evil.el]: You should mention drag support in the readme under Detailed Feature List / Mouse Interface.

**Comment by Alexander-Miller** [comment]
The commit message also needs the [Mouse] keyword.

File: src/elisp/treemacs-mouse-interface.el (hunk 3 above)
Comment (Compro-Prasad) [reply src/elisp/treemacs-mouse-interface.el]: So, what should be used instead?

File: src/elisp/treemacs-mouse-interface.el (hunk 4 above)
Comment (Compro-Prasad) [reply src/elisp/treemacs-mouse-interface.el]: Yeah. Will do that.

File: src/elisp/treemacs-mouse-interface.el (hunk 3 above)
Comment (Alexander-Miller) [reply src/elisp/treemacs-mouse-interface.el]: Nothing, just the the -current-button as is.

**Comment by Compro-Prasad** [comment]
I want to know how do you develop treemacs? I am currently having problem on how to select a specific git repository for developing a package. Spacemacs has this feature to use a git repo but for that I will have to change the actual code in spacemacs itself. Is there another way?

**Comment by Alexander-Miller** [comment]
I don't go through spacemacs. Treemacs and its subsidiaries are all part of the excluded-packages list. Isntead I load treemacs straight from the hdd via use package. See https://github.com/Alexander-Miller/dotfiles/blob/master/.config/spacemacs/user-config.org#treemacs

File: src/elisp/treemacs-mouse-interface.el (hunk 1 above)
Comment (Compro-Prasad) [reply src/elisp/treemacs-mouse-interface.el]: This has been resolved.

**Comment by Compro-Prasad** [comment]
Is there something I am missing?

**Comment by Alexander-Miller** [comment]
> Is there something I am missing?

Do you mean you're done with the requested changes? I think there should be a button somewhere for you to request a new review - that's what I was waiting for.

File: README.org [hunk 6]
```diff
@@ -176,6 +176,8 @@ A right click popup-menu is also available:
 
 [[file:screenshots/right-click.png]]
 
+You can also open a file in a specific window by dragging using left click from ~treemacs~ to the required window.
```
Comment (Compro-Prasad) [inline README.org]: This is the added comment in the README.

**Comment by Compro-Prasad** [comment]
I didn't know this feature of resolving conversations until you mentioned. I have resolved all o them from my side. Just test if this works.

File: README.org [hunk 7]
```diff
@@ -14,45 +14,45 @@
 - [[#quick-feature-overview][Quick Feature Overview]]
 - [[#quick-start][Quick Start]]
 - [[#detailed-feature-list][Detailed Feature List]]
-   - [[#projects-and-workspaces][Projects and Workspaces]]
```
Comment (Alexander-Miller) [inline README.org]: Try regenerating the TOC with `org-list-indent-offset` set to 1. That *should* fix all those indent changes.

**Comment by Compro-Prasad** [comment]
I have updated the TOC.

File: src/elisp/treemacs-macros.el [hunk 8]
```diff
@@ -134,7 +134,7 @@ Will bind to current button to 'btn' for the executon of the action forms.
 When NO-ERROR is non-nil no error will be thrown if no match for the button
 state is achieved."
   (declare (debug (&rest [sexp form])))
-  `(-if-let (btn (treemacs-current-button))
+  `(-if-let (btn treemacs--last-selected-button)
```
Comment (Alexander-Miller) [inline src/elisp/treemacs-macros.el]: This change breaks treemacs, ensuring that TAB only applies to the last clicked button. This whole thing seems to be more complicated than it first appeared. I think it's best if I take over from here.

Comment (Compro-Prasad) [reply src/elisp/treemacs-macros.el]: Sure, you know the code better than me. You can take over.

Comment (Alexander-Miller) [reply src/elisp/treemacs-macros.el]: Pushed my changes. Please make sure it works on your end as well, and I'll merge.

Comment (Compro-Prasad) [reply src/elisp/treemacs-macros.el]: Just tested it. Works.

**Comment by Alexander-Miller** [comment]
:+1:

**Pull request merged**